pkg/statistics/handle/usage/indexusage/collector.go

```diff
@@ -0,0 +1,218 @@
+// Copyright 2024 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package indexusage
+
+import (
+	"sync"
+	"time"
+
+	"github.com/pingcap/tidb/pkg/parser/model"
+)
+
+// GlobalIndexID is the key type for indexUsageMap.
+type GlobalIndexID struct {
+	TableID int64
+	IndexID int64
+}
+
+// Sample is the data structure to store index usage information.
+type Sample struct {
+	LastUsedAt     time.Time
+	QueryTotal     uint64
+	KvReqTotal     uint64
+	RowAccessTotal uint64
+	// 0, 0-1, 1-10, 10-20, 20-50, 50-100, 100
+	PercentageAccess [7]uint64
+}
+
+var bucketBound = [6]float64{0, 0.01, 0.1, 0.2, 0.5, 1.0}
+
+func getIndexUsageAccessBucket(percentage float64) int {
+	if percentage == 0 {
+		return 0
+	}
+
+	bucket := 0
+	for i := 1; i < len(bucketBound); i++ {
+		if percentage >= bucketBound[i-1] && percentage < bucketBound[i] {
+			bucket = i
+			break
+		}
+	}
+	if percentage == 1.0 {
+		bucket = len(bucketBound)
+	}
+
+	return bucket
+}
+
+// NewSample creates a new data point for index usage.
+func NewSample(queryTotal uint64, kvReqTotal uint64, rowAccess uint64, tableTotalRows uint64) *Sample {
```

Comment:
Should we maintain an object pool for `*Sample`? It seems like this allocation operation is very high-frequent.

Reply:
It's modified to use `Sample`.